Added a check to see if we are on the last block and see if any block…

gpotts opened the pull request:
There was a bug when outputting compression type "NM".  If an image segment had all blank tiles the offset was not being adjusted for the next image segment to the start of where the pixels would have been written.  I added a working variable to blockIO to specify the total blocks written and use this during the write to see when we are on the last tile.  If the last tile is blank and no other tiles have been written then we adjust the file offset to go past the Mask header and to the start of the pixel data.  This will now be pointing to the proper offset for the next image segment to be written.

JonathanMeans commented:
Would it be possible to include the test you wrote as well for our test suite?

gpotts commented:
Hello Jon:

I will attach the file to the request.   I basically added a sio output for each situation and then did a sio to nitf in order to simulate the cmv_ccd way of writing.   I just copied another test code to verify the NITF by printing the mask data.  Let me modify it to test if that particular segment is read in as an empty segment.

Take care

Garrett


From: JonathanMeans <notifications@github.com>
Reply-To: mdaus/nitro <reply@reply.github.com>
Date: Wednesday, March 6, 2019 at 2:47 PM
To: mdaus/nitro <nitro@noreply.github.com>
Cc: "Potts, Garrett" <Garrett.Potts@radiantsolutions.com>, Author <author@noreply.github.com>
Subject: Re: [mdaus/nitro] Added a check to see if we are on the last block and see if any block… (#120)


Would it be possible to include the test you wrote as well for our test suite?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/mdaus/nitro/pull/120#issuecomment-470249356>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACL9v4R0mmCMY51g_5ReQPuuyEcCtoK0ks5vUBsxgaJpZM4bhjb6>.

This electronic communication and any attachments may contain confidential and proprietary information of Radiant Solutions, Inc. If you are not the intended recipient, or an agent or employee responsible for delivering this communication to the intended recipient, or if you have received this communication in error, please do not print, copy, retransmit, disseminate or otherwise use the information. Please indicate to the sender that you have received this communication in error, and delete the copy you received.

Radiant Solutions reserves the right to monitor any electronic communication sent or received by its employees, agents or representatives.

gpotts commented:
[main.txt](https://github.com/mdaus/nitro/files/2938201/main.txt)
Please find the test code attached.  The only problem is that it includes a header from the six-library.  Not sure if this could be modified to be pure nitro only.

asylvest commented on modules/c/nitf/source/ImageIO.c:
```diff
@@ -7166,7 +7170,8 @@ NITFPRIV(int) nitf_ImageIO_writeToBlock(_nitf_ImageIOBlock * blockIO,
     }
-
+    nitf_Int64 nBlocks = nitf->nBlocksTotal;
+    nitf_Uint8 lastBlock = (blockIO->number == (nBlocks - 1));
```
Use `bool`

asylvest replied:
This is C so these need to be declared at the top of the function

asylvest commented on modules/c/nitf/source/ImageIO.c:
```diff
@@ -7194,6 +7199,17 @@ NITFPRIV(int) nitf_ImageIO_writeToBlock(_nitf_ImageIOBlock * blockIO,
             (*scanner)(blockIO, &padPresent, &dataPresent);
             if (!dataPresent)                  /* Pad only do not write */
             {
+                if (lastBlock&&(blockIO->totalBlocksWritten == 0))
```
Spaces before and after `&&`

asylvest commented on modules/c/nitf/source/ImageIO.c:
```diff
@@ -7266,7 +7282,7 @@ NITFPRIV(int) nitf_ImageIO_writeToBlock(_nitf_ImageIOBlock * blockIO,
                     blockIO->blockMask[blockIO->number];
         }
-
+        blockIO->totalBlocksWritten++;
```
Pre rather than post-increment this

gpotts commented:
Chan ge request was made.  Also,  an initial checking of the unit test was also made.  The unit test is called: test_image_segment_blank_nm_compression

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,331 @@
+                     
+}
+const int BLOCK_LENGTH = 1024;
+const int64_t ILOC_MAX = 99999;
```
For consistency (and to avoid C++11), can you use the NITF typedefs (`nitf::Int64`) instead of the `<ctypes.h>` types?

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,331 @@
+#include "TestCase.h"
+
+extern "C"{
+   NITF_BOOL nitf_ImageIO_getMaskInfo(nitf_ImageIO *nitf,
```
Style: with `*` here and `&` below, they should be on the type-side. (e.g. `nitf_ImageIO* nitf`)

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,331 @@
+                              size_t numCols)
+{
+   size_t numBlocksPresent = 0;
+   size_t row;
```
This is C++, so no need to declare everything up front.

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,331 @@
+                  const types::RowCol<size_t> &fullDims,
+                  const int segmentIdxToMakeEmpty)
+{
+   size_t bytes = fullDims.row * fullDims.col;
```
Can be const. (FYI, `RowCol` has an `area()` method)

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,331 @@
+   return numBlocksPresent;
+}
+void createSingleBandBuffer(std::vector<uint8_t>& buffer,                                     
+                  nitf::ImageSegmentComputer &segmentComputer,
```
Can be const

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,331 @@
+}
+
+void createBuffers(std::vector<std::vector<uint8_t> > &buffers,
+                   nitf::ImageSegmentComputer &imageSegmentComputer,
```
Can be const

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,331 @@
+   //
+   // Create 3 segments for testing
+   nitf_Error error; /* Pointer to the error object */
+   int64_t numberLines     = BLOCK_LENGTH * 3;
```
Const + nitf types here

asylvest commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,331 @@
+
+            /*  Get an image reader which creates the nitf_ImageIO were the masks are */
+
+            iReader = nitf_Reader_newImageReader(reader, imgCtr, NULL, &error);
```
Can't you use the C++ layer throughout here and then just use the getter methods to get the raw C pointer when you need it?

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,331 @@
+
+         in = nitf_IOHandle_create(tempNitf.pathname().c_str(),
+                                   NITF_ACCESS_READONLY, NITF_OPEN_EXISTING, &error);
+         TEST_ASSERT_TRUE( (NITF_INVALID_HANDLE(in)!=true) );
```
Use `TEST_ASSERT_FALSE` here

gpotts commented:
I had most as C++ and refactored but could not see a way to get the Mask via C++ interfaces.   All the other items I could do via C++ so I reverted back to C.




From: Adam Sylvester <notifications@github.com>
Reply-To: mdaus/nitro <reply@reply.github.com>
Date: Friday, March 8, 2019 at 7:16 AM
To: mdaus/nitro <nitro@noreply.github.com>
Cc: "Potts, Garrett" <Garrett.Potts@radiantsolutions.com>, Author <author@noreply.github.com>
Subject: Re: [mdaus/nitro] Added a check to see if we are on the last block and see if any block… (#120)


@asylvest requested changes on this pull request.

________________________________

In modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp<https://github.com/mdaus/nitro/pull/120#discussion_r263758101>:

> +         imgIter = nitf_List_begin(record->images);

+         imgEnd = nitf_List_end(record->images);

+         while (nitf_ListIterator_notEqualTo(&imgIter, &imgEnd))

+         {

+            /*  Get information from the image subheader */

+            seg = (nitf_ImageSegment *)nitf_ListIterator_get(&imgIter);

+            TEST_ASSERT_TRUE((seg!=NULL));

+            subhdr = seg->subheader;

+

+            nitf_Field_get(subhdr->imageMode,

+                           imageMode, NITF_CONV_STRING, NITF_IMODE_SZ + 1, &error);

+            imageMode[NITF_IMODE_SZ] = 0;

+

+            /*  Get an image reader which creates the nitf_ImageIO were the masks are */

+

+            iReader = nitf_Reader_newImageReader(reader, imgCtr, NULL, &error);

Can't you use the C++ layer throughout here and then just use the getter methods to get the raw C pointer when you need it?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/mdaus/nitro/pull/120#pullrequestreview-212258198>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACL9v_qSabRHdxP9qxfsNPrv5MaJulxmks5vUlSsgaJpZM4bhjb6>.

This electronic communication and any attachments may contain confidential and proprietary information of Radiant Solutions, Inc. If you are not the intended recipient, or an agent or employee responsible for delivering this communication to the intended recipient, or if you have received this communication in error, please do not print, copy, retransmit, disseminate or otherwise use the information. Please indicate to the sender that you have received this communication in error, and delete the copy you received.

Radiant Solutions reserves the right to monitor any electronic communication sent or received by its employees, agents or representatives.

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,331 @@
+
+int main(int argc, char *argv[])
+{
+   TEST_CHECK(testBlankSegmentsValid);
```
return 0;

gpotts commented:
I do believe all changes have been applied.  Will look again at the C++ implementation to see if I can find out how to access the Mask from the C++ interfaces.

gpotts commented:
Hello Adam:

I think I figured out how to use C++ instead for this one.  I will update and let y’all know


From: Adam Sylvester <notifications@github.com>
Reply-To: mdaus/nitro <reply@reply.github.com>
Date: Friday, March 8, 2019 at 7:16 AM
To: mdaus/nitro <nitro@noreply.github.com>
Cc: "Potts, Garrett" <Garrett.Potts@radiantsolutions.com>, Author <author@noreply.github.com>
Subject: Re: [mdaus/nitro] Added a check to see if we are on the last block and see if any block… (#120)


@asylvest requested changes on this pull request.

________________________________

In modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp<https://github.com/mdaus/nitro/pull/120#discussion_r263758101>:

> +         imgIter = nitf_List_begin(record->images);

+         imgEnd = nitf_List_end(record->images);

+         while (nitf_ListIterator_notEqualTo(&imgIter, &imgEnd))

+         {

+            /*  Get information from the image subheader */

+            seg = (nitf_ImageSegment *)nitf_ListIterator_get(&imgIter);

+            TEST_ASSERT_TRUE((seg!=NULL));

+            subhdr = seg->subheader;

+

+            nitf_Field_get(subhdr->imageMode,

+                           imageMode, NITF_CONV_STRING, NITF_IMODE_SZ + 1, &error);

+            imageMode[NITF_IMODE_SZ] = 0;

+

+            /*  Get an image reader which creates the nitf_ImageIO were the masks are */

+

+            iReader = nitf_Reader_newImageReader(reader, imgCtr, NULL, &error);

Can't you use the C++ layer throughout here and then just use the getter methods to get the raw C pointer when you need it?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/mdaus/nitro/pull/120#pullrequestreview-212258198>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACL9v_qSabRHdxP9qxfsNPrv5MaJulxmks5vUlSsgaJpZM4bhjb6>.

This electronic communication and any attachments may contain confidential and proprietary information of Radiant Solutions, Inc. If you are not the intended recipient, or an agent or employee responsible for delivering this communication to the intended recipient, or if you have received this communication in error, please do not print, copy, retransmit, disseminate or otherwise use the information. Please indicate to the sender that you have received this communication in error, and delete the copy you received.

Radiant Solutions reserves the right to monitor any electronic communication sent or received by its employees, agents or representatives.

gpotts commented:
Hello All:

I was able to refactor to use C++ instead of C.  There was only one C API call that had to be made and the was to get the Mask Information

Take care

Garrett

From: Adam Sylvester <notifications@github.com>
Reply-To: mdaus/nitro <reply@reply.github.com>
Date: Friday, March 8, 2019 at 7:16 AM
To: mdaus/nitro <nitro@noreply.github.com>
Cc: "Potts, Garrett" <Garrett.Potts@radiantsolutions.com>, Author <author@noreply.github.com>
Subject: Re: [mdaus/nitro] Added a check to see if we are on the last block and see if any block… (#120)


@asylvest requested changes on this pull request.

________________________________

In modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp<https://github.com/mdaus/nitro/pull/120#discussion_r263758101>:

> +         imgIter = nitf_List_begin(record->images);

+         imgEnd = nitf_List_end(record->images);

+         while (nitf_ListIterator_notEqualTo(&imgIter, &imgEnd))

+         {

+            /*  Get information from the image subheader */

+            seg = (nitf_ImageSegment *)nitf_ListIterator_get(&imgIter);

+            TEST_ASSERT_TRUE((seg!=NULL));

+            subhdr = seg->subheader;

+

+            nitf_Field_get(subhdr->imageMode,

+                           imageMode, NITF_CONV_STRING, NITF_IMODE_SZ + 1, &error);

+            imageMode[NITF_IMODE_SZ] = 0;

+

+            /*  Get an image reader which creates the nitf_ImageIO were the masks are */

+

+            iReader = nitf_Reader_newImageReader(reader, imgCtr, NULL, &error);

Can't you use the C++ layer throughout here and then just use the getter methods to get the raw C pointer when you need it?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/mdaus/nitro/pull/120#pullrequestreview-212258198>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACL9v_qSabRHdxP9qxfsNPrv5MaJulxmks5vUlSsgaJpZM4bhjb6>.

This electronic communication and any attachments may contain confidential and proprietary information of Radiant Solutions, Inc. If you are not the intended recipient, or an agent or employee responsible for delivering this communication to the intended recipient, or if you have received this communication in error, please do not print, copy, retransmit, disseminate or otherwise use the information. Please indicate to the sender that you have received this communication in error, and delete the copy you received.

Radiant Solutions reserves the right to monitor any electronic communication sent or received by its employees, agents or representatives.

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,296 @@
+         {
+            nitf::ImageSegment seg(static_cast<nitf_ImageSegment *>((*imageIterator)));
+            nitf::ImageSubheader subhdr     = seg.getSubheader();
+            std::string imageMode           = subhdr.getImageMode().toString();
```
Does this get used anywhere?

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,296 @@
+            }
+            ++imgCtr;
+         }
+         input_io.close();
```
I believe the destructor actually handles this for you.

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,296 @@
+          '\0', segmentSizeInBytes);
+}
+
+void createBuffers(std::vector<std::vector<nitf::Uint8> > &buffers,
```
And then make sure the references get pushed to the left too, whenever they pop up. (`std::vector<nitf::Uint8> >& buffers`)

JonathanMeans commented on modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp:
```diff
@@ -0,0 +1,296 @@
+         nitf_Uint64 *padMask=NULL;               /* Pad mask array */
+         nitf::Int64 imgCtr = 0;
+         /*  Image information */
+         // char imageMode[NITF_IMODE_SZ + 1]; /* Image (blocking) mode */
```
Can this comment be deleted?

gpotts commented:
Applied modifications

From: JonathanMeans <notifications@github.com>
Reply-To: mdaus/nitro <reply@reply.github.com>
Date: Friday, March 8, 2019 at 10:19 AM
To: mdaus/nitro <nitro@noreply.github.com>
Cc: "Potts, Garrett" <Garrett.Potts@radiantsolutions.com>, Author <author@noreply.github.com>
Subject: Re: [mdaus/nitro] Added a check to see if we are on the last block and see if any block… (#120)


@JonathanMeans requested changes on this pull request.

________________________________

In modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp<https://github.com/mdaus/nitro/pull/120#discussion_r263811288>:

> +         nitf::Int64 imgCtr = 0;

+         /*  Image information */

+         // char imageMode[NITF_IMODE_SZ + 1]; /* Image (blocking) mode */

+         nitf::IOHandle input_io(tempNitf.pathname(),

+                                 NITF_ACCESS_READONLY,

+                                 NITF_OPEN_EXISTING);

+         nitf::Reader reader;

+         nitf::Record record = reader.read(input_io);

+         nitf::List images   = record.getImages();

+         for (nitf::ListIterator imageIterator = images.begin();

+              imageIterator != images.end();

+              ++imageIterator)

+         {

+            nitf::ImageSegment seg(static_cast<nitf_ImageSegment *>((*imageIterator)));

+            nitf::ImageSubheader subhdr     = seg.getSubheader();

+            std::string imageMode           = subhdr.getImageMode().toString();

Does this get used anywhere?

________________________________

In modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp<https://github.com/mdaus/nitro/pull/120#discussion_r263812014>:

> +

+            const nitf::Int64 nBlocksPresent = getNumberBlocksPresent(blockMask,

+                                                                      blockingInfo.getNumBlocksPerRow(),

+                                                                      blockingInfo.getNumBlocksPerCol());

+

+            if (imgCtr == testIdx)

+            {

+               TEST_ASSERT_EQ(nBlocksPresent, 0);

+            }

+            else

+            {

+               TEST_ASSERT_EQ(nBlocksPresent, totalBlocks);

+            }

+            ++imgCtr;

+         }

+         input_io.close();

I believe the destructor actually handles this for you.

________________________________

In modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp<https://github.com/mdaus/nitro/pull/120#discussion_r263812441>:

> +                            const nitf::ImageSegmentComputer &segmentComputer,

+                            const types::RowCol<nitf::Int64> &fullDims,

+                            const nitf::Int64 segmentIdxToMakeEmpty)

+{

+   const nitf::Int64 bytes = fullDims.area();

+   const std::vector<nitf::ImageSegmentComputer::Segment> &segments = segmentComputer.getSegments();

+   /* All segments should be the same size in this test so this is safe */

+   const nitf::Int64 segmentSizeInBytes = segments[segmentIdxToMakeEmpty].numRows * fullDims.col;

+

+   buffer.resize(bytes);

+   memset(&buffer.front(), '0xff', bytes);

+   memset(&buffer.front() + (segmentSizeInBytes * segmentIdxToMakeEmpty),

+          '\0', segmentSizeInBytes);

+}

+

+void createBuffers(std::vector<std::vector<nitf::Uint8> > &buffers,

And then make sure the references get pushed to the left too, whenever they pop up. (std::vector<nitf::Uint8> >& buffers)

________________________________

In modules/c++/nitf/unittests/test_image_segment_blank_nm_compression.cpp<https://github.com/mdaus/nitro/pull/120#discussion_r263812988>:

> +         imageWriter.attachSource(iSource);

+      }

+      writer.write();

+      output_io.close();

+

+      {

+         nitf_Uint32 imageDataOffset=0;        /* Offset to actual image data past masks */

+         nitf_Uint32 blockRecordLength=0;      /* Block mask record length */

+         nitf_Uint32 padRecordLength=0;        /* Pad mask record length */

+         nitf_Uint32 padPixelValueLength=0;    /* Pad pixel value length in bytes */

+         nitf_Uint8 *padValue = NULL;        /* Pad value */

+         nitf_Uint64 *blockMask=NULL;             /* Block mask array */

+         nitf_Uint64 *padMask=NULL;               /* Pad mask array */

+         nitf::Int64 imgCtr = 0;

+         /*  Image information */

+         // char imageMode[NITF_IMODE_SZ + 1]; /* Image (blocking) mode */

Can this comment be deleted?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/mdaus/nitro/pull/120#pullrequestreview-212326032>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACL9v5sCPInFQXQmIStOT23_Q_5oTCz2ks5vUn8pgaJpZM4bhjb6>.

This electronic communication and any attachments may contain confidential and proprietary information of Radiant Solutions, Inc. If you are not the intended recipient, or an agent or employee responsible for delivering this communication to the intended recipient, or if you have received this communication in error, please do not print, copy, retransmit, disseminate or otherwise use the information. Please indicate to the sender that you have received this communication in error, and delete the copy you received.

Radiant Solutions reserves the right to monitor any electronic communication sent or received by its employees, agents or representatives.
